faiss/impl/io.cpp

```diff
@@ -55,6 +55,28 @@ size_t VectorIOReader::operator()(void* ptr, size_t size, size_t nitems) {
     return nitems;
 }
 
+/***********************************************************************
+ * IO Buffer
+ ***********************************************************************/
+
+size_t BufIOReader::operator()(void* ptr, size_t size, size_t nitems) {
+    if (rp >= buf_size)
+        return 0;
+    size_t nremain = (buf_size - rp) / size;
+    if (nremain < nitems)
+        nitems = nremain;
+    if (size * nitems > 0) {
+        memcpy(ptr, &buf[rp], size * nitems);
+        rp += size * nitems;
+    }
+
+    return nitems;
+}
+
+BufIOReader::~BufIOReader() {
```

Comment:
Shouldn't we be calling `free()` over buf here?

Reply:
the main reason why i avoided that is because shifting the responsibility to the caller helps making better decisions as to whether we want to reuse the pointer and the memory block (which i've commented on the go-faiss write-read path) thereby reducing the costs of allocation or just free the memory block of that pointer (which we are doing right now and is being revisited for reusability purposes). please let me know if that makes sense, thanks

Reply:
Ok, we want to optimize more down this stack in the C++ side of things. We'll chat next week.